core/src/replay_stage.rs

```diff
@@ -850,34 +850,50 @@ impl ReplayStage {
         }
     }
 
+    fn first_of_consecutive_leader_slots(slot: Slot) -> Slot {
+        (slot / NUM_CONSECUTIVE_LEADER_SLOTS) * NUM_CONSECUTIVE_LEADER_SLOTS
+    }
+
     fn retransmit_latest_unpropagated_leader_slot(
         poh_recorder: &Arc<Mutex<PohRecorder>>,
         retransmit_slots_sender: &RetransmitSlotsSender,
         progress: &mut ProgressMap,
     ) {
         let start_slot = poh_recorder.lock().unwrap().start_slot();
-        if let Some(latest_leader_slot) = progress.get_latest_leader_slot(start_slot) {
-            if !progress
-                .get_propagated_stats(latest_leader_slot)
-                .map(|stats| stats.is_propagated)
-                .unwrap_or(true)
-            {
-                warn!("Slot not propagated: slot={}", latest_leader_slot);
-                let retransmit_info = progress.get_retransmit_info(latest_leader_slot).unwrap();
-                if retransmit_info.reached_retransmit_threshold() {
-                    info!(
-                        "Retrying retransmit: start_slot={} latest_leader_slot={} retransmit_info={:?}",
-                        start_slot, latest_leader_slot, &retransmit_info,
-                    );
-                    datapoint_info!(
-                        "replay_stage-retransmit-timing-based",
-                        ("slot", latest_leader_slot, i64),
-                        ("retry_iteration", retransmit_info.retry_iteration, i64),
-                    );
-                    let _ = retransmit_slots_sender.send(latest_leader_slot);
-                    retransmit_info.increment_retry_iteration();
-                } else {
-                    info!("Bypass retry: {:?}", &retransmit_info);
+        if let (false, Some(latest_leader_slot)) =
+            progress.get_leader_propagation_slot_must_exist(start_slot)
+        {
+            info!(
+                "Slot not propagated: start_slot={} latest_leader_slot={}",
+                start_slot, latest_leader_slot
+            );
+
+            let first_leader_group_slot =
+                Self::first_of_consecutive_leader_slots(latest_leader_slot);
+
+            for slot in first_leader_group_slot..=latest_leader_slot {
+                if !progress.is_propagated(slot).unwrap_or(true) {
+                    if let Some(retransmit_info) = progress.get_retransmit_info(slot) {
+                        if retransmit_info.reached_retransmit_threshold() {
+                            info!(
```

Comment:
is it possible to extract this check into a common function, that can also be called to replace the similar logic in `maybe_start_leader()`?

Reply:
done